egs/madcat_ar/v1/run_end2end.sh

```diff
@@ -74,8 +74,15 @@ fi
 
 if [ $stage -le 5 ]; then
   echo "$0: Preparing dictionary and lang..."
+  cut -d' ' -f2- data/train/text | python3 local/reverse.py | python3 local/prepend_words.py | python3 local/learn_bpe.py -s 700 > data/train/bpe.out
+  for set in test train dev ; do
+    cut -d' ' -f1 data/$set/text > data/$set/ids
+    cut -d' ' -f2- data/$set/text | python3 local/reverse.py | python3 local/prepend_words.py | python3 local/apply_bpe.py -c data/train/bpe.out | sed 's/@@//g' > data/$set/bpe_text
+    mv data/$set/text data/$set/text.old
+    paste -d' ' data/$set/ids data/$set/bpe_text > data/$set/text
+  done
   local/prepare_dict.sh
-  utils/prepare_lang.sh --num-sil-states 4 --num-nonsil-states 8 --sil-prob 0.9999 \
+  utils/prepare_lang.sh --num-sil-states 4 --num-nonsil-states 8 --sil-prob 0.0 --position-dependent-phones false \
```

Comment:
you might want to comment that because this recipe uses BPE encoding, the silences are part of the words' pronunciations.

This recipe is not 100% correct in terms of how the word-final silence of the sentence is treated.  Basically, our model won't allow it as an option.  It can probably be handled by the context dependency but it's not ideal (e.g. it will mess up the alignments, if we cared about that) and I think it makes sense to fix it.

Please try the following fix.   Add a script in a shared location (e.g.  steps/bpe/ if that's what we want to do)

steps/bpe/add_final_optional_silence.sh


which will have usage:
```
if [ $# -ne 1 ]; then
  echo "Usage: $0  <lang> "
  echo " Add final optional silence to lexicon FSTs (L.fst an L_disambig.fst) in  "
  echo " lang/ directory <lang>."
fi
```
It can use "fstconcat" for this, concatenating with an FST like this:
```
0  1   $opt_sil_phone 0  -0.69
0 -0.69
1  0.0
```
The -0.69 is log(0.5) so it has equal probability for silence and not-silence.  This is all based on the assumption that your BPE puts the | at the beginning of words and the silence is at the beginning of words.   However, you should probably make the silence prob configurable via the --sil-prob option like prepare_lang.sh.  Be sure to treat the special case sil_prob=1.0 specially (disallow sil_prob=0.0, in that we just shouldn't call the script).

We can consider making the final silence non-optional, i.e. prob=1.0, since the initial silence is effectively non-optional, given how the BPE works.  You can test that, and also the topologies e.g. the num-states in silence.

Reply:
I will test all of these strategies, but I think I should get some help from @hhadian to do them right. Because I don't have much experience in working with FSTs. 

Should I test BPE on IAM with this kind of recipe or wait to fix the final-word space issue?

Reply:
@danpovey, what would you think about adding a `SIL` at the end of the transcriptions (i.e. in `text`) given the fact that we always have padding for OCR recipes (so we always have a final silence)?